### PR TITLE
Support for vrm.

### PR DESCRIPTION
**Pull request by fire**
Use a vrm file extension to trigger processing.

See #198 for test cases and examples.

The command to test is now:

`./gltfpack -i "Miraikomachi.vrm" -o "Miraikomachi Opt.VRM" `

### CODE REVIEW

**Comment by fire** [comment]
How do I read those errors in the cicd?

**Comment by fire** [comment]
@zeux I wasn't able to recreate the errors. Do you know how that test works?

**Comment by zeux** [comment]
The test validates that packing any model from glTF-Sample-Models repository produces a valid glTF file as validated by glTF-Validator. glTF-Sample-Models/2.0/Cameras/glTF/Cameras.gltf is the specific file on which the process fails with this change, you should be able to get the source file from glTF-Sample-Models Khronos repository.

**Comment by fire** [comment]
Green!

File: gltf/parsegltf.cpp [hunk 1]
```diff
@@ -165,6 +175,8 @@ static void parseMeshesGltf(cgltf_data* data, std::vector<Mesh>& meshes, std::ve
 			meshes.push_back(Mesh());
 			Mesh& result = meshes.back();
 
+			result.name = mesh.name;
```
Comment (fire) [inline gltf/parsegltf.cpp:178]: Might be a problem with variable lifetime

File: gltf/gltfpack.cpp [hunk 2]
```diff
@@ -488,6 +498,11 @@ static void process(cgltf_data* data, const char* input_path, const char* output
 				append(json_meshes, ",\"indices\":");
 				append(json_meshes, index_accr);
 			}
+			if (settings.keep_extras && !prim.extras.empty())
```
Comment (zeux) [inline gltf/gltfpack.cpp:501]: Why is this necessary? AFAICT the only "extras" node in the .VRM file from the issue is on the mesh, not on primitive, and that contains targetNames which should already be preserved by gltfpack.

File: gltf/gltfpack.cpp [hunk 3]
```diff
@@ -661,10 +678,39 @@ static void process(cgltf_data* data, const char* input_path, const char* output
...
+	append(json_extensions, "\"extensions\":{");
+	if (settings.vrm)
+	{
+
+		std::string extension_vrm;
+		extension_vrm.assign("VRM");
+		for (size_t i = 0; i < data->data_extensions_count; ++i)
+		{
+			cgltf_extension extension = data->data_extensions[i];
+			if (!extension.name)
```
Comment (zeux) [inline gltf/gltfpack.cpp:696]: This & above & below can be replaced with `if (strcmp(extension.name, "VRM") == 0)` (extension.name is guaranteed to be non-null)

File: gltf/gltfpack.cpp [hunk 4]
```diff
@@ -676,6 +722,8 @@ static void process(cgltf_data* data, const char* input_path, const char* output
 	writeArray(json, "meshes", json_meshes);
 	writeArray(json, "skins", json_skins);
 	writeArray(json, "animations", json_animations);
+	comma(json);
+	append(json, json_extensions);
```
Comment (zeux) [inline gltf/gltfpack.cpp:726]: This will always emit "extensions" and also causes a conflict with top-level KHR_lights_punctual extension since there's now going to be two extension blocks for files that use that.

Instead what should happen is:
- `json_extensions` should be initially empty
- `json_lights` should be appended to it, if necessary
- VRM data should be appended to it, if necessary
- if `json_extensions` isn't empty when forming the final JSON blob, we should write out `"extensions":{` followed by json_extensions contents followed by `}`

File: gltf/gltfpack.cpp [hunk 5]
```diff
@@ -661,10 +678,39 @@ static void process(cgltf_data* data, const char* input_path, const char* output
...
+			std::string extension_tested;
+			extension_tested.assign(extension.name);
+			if (extension_vrm != extension_tested)
+			{
+				break;
+			}
+			append(json_extensions, "\"");
+			append(json_extensions, extension.name);
+			append(json_extensions, "\":");
+			append(json_extensions, extension.data);
```
Comment (zeux) [inline gltf/gltfpack.cpp:709]: This should use `appendJson` to minify the output

File: gltf/gltfpack.cpp [hunk 6]
```diff
@@ -790,8 +838,12 @@ int gltfpack(const char* input, const char* output, const char* report, Settings
 
 	const char* iext = strrchr(input, '.');
 	const char* oext = output ? strrchr(output, '.') : NULL;
+	
+	if (iext && (strcmp(iext, ".vrm") == 0 || strcmp(iext, ".VRM") == 0)) {
```
Comment (zeux) [inline gltf/gltfpack.cpp:842]: Please use clang-format on this code, as this bracing style is inconsistent with the coding guidelines, see https://github.com/zeux/meshoptimizer/blob/master/CONTRIBUTING.md

File: gltf/gltfpack.cpp [hunk 7]
```diff
@@ -843,6 +895,16 @@ int gltfpack(const char* input, const char* output, const char* report, Settings
 		settings.texture_embed = true;
 	}
 
+	if (oext && (strcmp(oext, ".vrm") == 0 || strcmp(oext, ".VRM") == 0))
+	{
+		settings.vrm = true;
+		settings.keep_extras = true;
```
Comment (zeux) [inline gltf/gltfpack.cpp:901]: I would prefer to omit these, esp. `quantize` but preferably others. It's not clear to me why these need to be present, and if a user wants to apply these optimizations to a VRM file there's not going to be a way to do this without adding more kinds of command-line options, if `vrm` extension acts as an override.

File: gltf/gltfpack.cpp [hunk 8]
```diff
@@ -896,7 +958,7 @@ int gltfpack(const char* input, const char* output, const char* report, Settings
 			return 4;
 		}
 	}
-	else if (oext && (strcmp(oext, ".glb") == 0 || strcmp(oext, ".GLB") == 0))
+	else if (oext && (strcmp(oext, ".vrm") == 0 || strcmp(oext, ".VRM") == 0 || strcmp(oext, ".glb") == 0 || strcmp(oext, ".GLB") == 0))
```
Comment (zeux) [inline gltf/gltfpack.cpp:961]: Should this also be added to the condition above that sets `texture_embed`? Unclear what the de-facto rules are for VRM files, e.g. whether they are supposed to keep textures separately or not.

File: gltf/gltfpack.h [hunk 9]
```diff
@@ -39,6 +39,7 @@ struct Transform
 
 struct Mesh
 {
+	const char* name;
```
Comment (zeux) [inline gltf/gltfpack.h:42]: `std::string` might be better.

File: gltf/write.cpp [hunk 10]
```diff
@@ -739,6 +739,14 @@ static std::string decodeUri(const char* uri)
 
 void writeImage(std::string& json, std::vector<BufferView>& views, const cgltf_image& image, const ImageInfo& info, size_t index, const char* input_path, const char* output_path, const Settings& settings)
 {
+	if (image.name)
```
Comment (zeux) [inline gltf/write.cpp:742]: Is preserving the name of the image object necessary? It doesn't carry semantical meaning, and gltfpack right now only preserves names for objects that are likely to be used by-name externally, specifically nodes, materials and animations.

File: gltf/write.cpp [hunk 11]
```diff
@@ -1320,7 +1328,8 @@ void writeAnimation(std::string& json, std::vector<BufferView>& views, std::stri
 
 		comma(json_samplers);
 		append(json_samplers, "{\"input\":");
-		append(json_samplers, range ? range_accr : track.constant ? pose_accr : time_accr);
+		append(json_samplers, range ? range_accr : track.constant ? pose_accr
```
Comment (zeux) [inline gltf/write.cpp:1331]: Was this change intended?

**Comment by fire** [comment]
Thank you for the review. I will be delayed until the weekend to submit the changes.

File: gltf/write.cpp (hunk 11 above)
Comment (fire) [reply gltf/write.cpp:1331]: Not intended.

File: gltf/write.cpp (hunk 10 above)
Comment (fire) [reply gltf/write.cpp:742]: I need to check if this is required by the internal material definition called MTOON.

**Comment by fire** [comment]
I slept thinking about this optimization, because Godot Engine added autolod, I don't have a requirement for this. This VRM pr is salvageable.

**Comment by arpu** [comment]
still interessded in vrm optimation

@fire  can you point me to how Godot Engine autolod works?

**Comment by fire** [comment]
1. Godot load gltf2
2. For every imported mesh run mesh optimizer with the error returned
3. (Would like tesselate and the decimate option) because simpify gets stucks
4. Implement vrm

I believe that mesh optimizer works under wasm, so you can follow the same flow.
